File: src-backend/routes/webhook.js
const express = require('express');
const router = express.Router();
const { supabase } = require('../db/client');
const { handleIncomingMessage } = require('../../agents/orchestrator');

const VERIFY_TOKEN = process.env.META_VERIFY_TOKEN || 'change_me';

const normalizeResponseText = (text) => {
  if (!text) return '';
  return text.toString().trim().toUpperCase().replace(/[^A-ZÁÉÍÓÚÑ]/g, '');
};

// 1. Endpoint de verificación de Meta (GET) - Se mantiene igual
router.get('/', (req, res) => {
  const mode = req.query['hub.mode'];
  const token = req.query['hub.verify_token'];
  const challenge = req.query['hub.challenge'];

  if (mode && token) {
    if (mode === 'subscribe' && token === VERIFY_TOKEN) {
      return res.status(200).send(challenge);
    } else {
      return res.sendStatus(403);
    }
  }
  res.sendStatus(400);
});

// 2. Recepción de eventos del Webhook (POST) - Migrado a Supabase
router.post('/', async (req, res) => {
  const body = req.body;

  try {
    // Guardar el log del evento crudo en Supabase
    const { error: logError } = await supabase
      .from('webhook_events')
      .insert([{ raw_json: body, type: 'meta_webhook' }]);

    if (logError) console.error('Error insertando webhook_event:', logError);

    const entries = body.entry || [];
    
    // Iteramos sobre los datos recibidos
    for (const entry of entries) {
      const changes = entry.changes || [];
      for (const change of changes) {
        const messages = change.value?.messages || [];
        
        for (const message of messages) {
          const from = message.from;
          const text = message.text?.body || message.button?.text || message.interactive?.button_reply?.title || message.interactive?.list_reply?.title;
          
          // Si es un mensaje de texto entrante y no una actualización de estado
          if (text && !message.status) {
            handleIncomingMessage(from, text, 'whatsapp')
              .catch(err => console.error('Error en el agente del webhook handler:', err));
          }

          // Validación de palabras clave para confirmación automática de citas
          const normalized = normalizeResponseText(text);
          const confirmationKeywords = ['SI', 'SÍ', 'CONFIRMAR', 'ACEPTAR'];
          
          if (confirmationKeywords.includes(normalized)) {
            // Buscamos el contacto por teléfono en Supabase
            const { data: contact, error: contactError } = await supabase
              .from('contacts')
              .select('id')
              .eq('phone', from)
              .maybeSingle();

            if (!contactError && contact) {
              // Actualizamos el estado de la cita programada (scheduled) a confirmada (confirmed)
              await supabase
                .from('appointments')
                .update({ status: 'confirmed', updated_at: new Date().toISOString() })
                .eq('contact_id', contact.id)
                .eq('status', 'scheduled');
            }
          }

          // Si el webhook reporta un cambio de estado del mensaje enviado (sent, delivered, read)
          if (message.status) {
            await supabase
              .from('messages')
              .update({ status: message.status })
              .eq('meta_message_id', message.id);
          }
        }
      }
    }

    // Respondemos inmediatamente a Meta con un 200 OK para evitar reintentos duplicados
    res.sendStatus(200);

  } catch (err) {
    console.error('Error general procesando webhook event:', err);
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: src-backend/routes/webhook.py
import os
import json
import logging
from fastapi import APIRouter, Request, Response, BackgroundTasks, HTTPException, status
from db.client import db_client
from agents.orchestrator import handle_incoming_message

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/webhook", tags=["webhook"])

VERIFY_TOKEN = os.getenv("META_VERIFY_TOKEN", "change_me")

def normalize_response_text(text: str) -> str:
    """
    Normaliza el texto del mensaje para simplificar la comparación de palabras clave.
    """
    if not text:
        return ""
    s = text.strip().upper()
    return "".join(c for c in s if c.isalnum() or c == " ")

async def process_webhook_payload(body: dict):
    """
    Procesador del Webhook en segundo plano (Background Task).
    Garantiza respuestas de baja latencia a Meta y previene problemas de timeout.
    """
    try:
        # 1. Guardar evento de webhook
        await db_client.run(
            "INSERT INTO webhook_events (raw_json, type, received_at) VALUES (?, 'meta_webhook', datetime('now'))",
            (json.dumps(body),)
        )
        
        entries = body.get("entry", [])
        for entry in entries:
            changes = entry.get("changes", [])
            for change in changes:
                value = change.get("value", {})
                messages = value.get("messages", [])
                
                for message in messages:
                    sender_id = message.get("from")
                    
                    # Extraer el texto
                    text = (
                        message.get("text", {}).get("body", "") or 
                        message.get("button", {}).get("text", "") or 
                        message.get("interactive", {}).get("button_reply", {}).get("title", "") or 
                        message.get("interactive", {}).get("list_reply", {}).get("title", "")
                    )
                    
                    message_status = message.get("status")
                    message_id = message.get("id")
                    
                    # Manejar reportes de estado (entregado, leído, etc.)
                    if message_status and message_id:
                        await db_client.run(
                            "UPDATE messages SET status = ?, last_attempt_at = datetime('now') WHERE meta_message_id = ?",
                            (message_status, message_id)
                        )
                        continue
                        
                    # Procesar nuevo mensaje recibido del cliente
                    if text and not message_status:
                        # Iniciar orquestador de bot asíncronamente
                        await handle_incoming_message(sender_id, text, "whatsapp")
                        
                        # Validar palabras de confirmación de cita
                        normalized = normalize_response_text(text)
                        confirmation_keywords = ["SI", "SÍ", "CONFIRMAR", "ACEPTAR"]
                        if any(kw in normalized for kw in confirmation_keywords):
                            contact = await db_client.get("SELECT id FROM contacts WHERE phone = ?", (sender_id,))
                            if contact:
                                await db_client.run(
                                    """
                                    UPDATE appointments 
                                    SET status = 'confirmed', updated_at = datetime('now') 
                                    WHERE contact_id = ? AND status = 'scheduled'
                                    """,
                                    (contact["id"],)
                                )
    except Exception as e:
        logger.error(f"Error procesando webhook de Meta en segundo plano: {e}")

@router.get("")
async def verify_webhook(request: Request):
    """
    Endpoint de verificación GET requerido por Meta para validar el webhook.
    """
    params = request.query_params
    mode = params.get("hub.mode")
    token = params.get("hub.verify_token")
    challenge = params.get("hub.challenge")
    
    if mode and token:
        if mode == "subscribe" and token == VERIFY_TOKEN:
            logger.info("✅ Webhook verificado correctamente por Meta")
            return Response(content=challenge, media_type="text/plain")
        else:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Verificación fallida")
            
    raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Solicitud incorrecta")

@router.post("")
async def receive_webhook(request: Request, background_tasks: BackgroundTasks):
    """
    Receptor de eventos del webhook de Meta (POST).
    Retorna 200 OK inmediatamente y delega el procesamiento a BackgroundTasks.
    """
    try:
        body = await request.json()
    except Exception:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="JSON no válido")
        
    # Encolar la tarea en segundo plano no bloqueante
    background_tasks.add_task(process_webhook_payload, body)
    return {"status": "ok"}
